src/java.base/share/classes/javax/crypto/spec/HKDFParameterSpec.java

```diff
@@ -0,0 +1,493 @@
+            return salts;
+        }
+
+    }
+
+    /**
+     * Defines the input parameters of an Expand operation as defined in <a
+     * href="http://tools.ietf.org/html/rfc5869">RFC 5869</a>.
+     */
+    @PreviewFeature(feature = PreviewFeature.Feature.KEY_DERIVATION)
+    final class Expand implements HKDFParameterSpec {
+
+        // HKDF-Expand(PRK, info, L) -> OKM
+        private final SecretKey prk;
+        private final byte[] info;
+        private final int length;
+
+        /**
+         * Constructor that may be used to initialize an {@code Expand} object
+         *
+         * @param prk
+         *     the pseudorandom key; may be {@code null}
+         * @param info
+         *     the optional context and application specific information (may be
+         *     {@code null}); the byte[] is copied to prevent subsequent
+         *     modification
+         * @param length
+         *     the length of the output key material (must be > 0 and < 255 *
+         *     HMAC length)
+         *
+         * @throws IllegalArgumentException
+         *     if {@code length} not > 0
+         */
+        private Expand(SecretKey prk, byte[] info, int length) {
+            // a null prk could be indicative of ExtractThenExpand
+            this.prk = prk;
+            this.info = (info == null) ? null : info.clone();
+            if (!(length > 0)) {
+                throw new IllegalArgumentException("length must be > 0");
+            }
+            this.length = length;
+        }
+
+        /**
+         * Returns the pseudorandom key.
+         *
+         * @return the pseudorandom key
+         */
+        public SecretKey prk() {
+            return prk;
+        }
+
+        /**
+         * Returns the optional context and application specific information.
+         *
+         * @return a copy of the optional context and application specific
+         *     information, or {@code null} if not specified
+         */
+        public byte[] info() {
+            return (info == null) ? null : info.clone();
+        }
+
+        /**
+         * Returns the length of the output key material.
+         *
+         * @return the length of the output key material
+         */
+        public int length() {
+            return length;
+        }
+
+    }
+
+    /**
+     * Defines the input parameters of an ExtractThenExpand operation as defined
+     * in <a href="http://tools.ietf.org/html/rfc5869">RFC 5869</a>.
+     */
+    @PreviewFeature(feature = PreviewFeature.Feature.KEY_DERIVATION)
+    final class ExtractThenExpand implements HKDFParameterSpec {
+        private final Extract ext;
+        private final Expand exp;
+
+        /**
+         * Constructor that may be used to initialize an
+         * {@code ExtractThenExpand} object
+         * <p>
+         * Note: {@code addIKMValue} and {@code addSaltValue} may be called
+         * afterward to supply additional values, if desired
```

Comment:
What does this mean? {@code addIKMValue} and {@code addSaltValue} are methods of (@code Builder} class and do not belong to the {@code ExtractThenExpand} class. Copy-n-paste error?

Reply:
Yes, resolved.